### ISSUE TITLE
Full width layout support

### ISSUE THREAD

**Issue by davidsherlock** [easydigitaldownloads/themedd#16]
It would be useful within the customizer to have the ability to enable a full-width layout mode for the theme. Just a suggestion :)

**Comment by sumobi** [comment]
@davidsherlock Absolutely 👍

**Comment by amdrew** [comment]
New setting for this under `Theme Options` in the customizer.

**Comment by davidsherlock** [comment]
@amdrew, I might be missing something here but with the recent commit my site still looks the exact same (cache was flushed):

![screenshot 2017-06-17 11 57 34](https://user-images.githubusercontent.com/22668877/27252354-489d1c22-5354-11e7-8aaf-ac6c1fce838f.png)

**Comment by davidsherlock** [comment]
One more related bug report regarding the full-width mode, I forced the full-width mode via CSS, but when active the header image size needs to disable cropping in the customizer to allow a full width image to display.

![screenshot 2017-06-17 15 31 02](https://user-images.githubusercontent.com/22668877/27253741-fea4b15c-5371-11e7-95d0-1157dc2d07ce.png)
 full-width mode needs to

**Comment by amdrew** [comment]
Ok, full-width mode should really work now.

> but when active the header image size needs to disable cropping in the customizer to allow a full width image to display.

@davidsherlock How does one do that?

As far as I'm aware the image generated as the custom header follows the width set in the `add_theme_support( 'custom-header' )` call (https://codex.wordpress.org/Custom_Headers). 

In Themedd, that's set to 1480 by default: https://github.com/easydigitaldownloads/themedd/blob/master/includes/customizer.php#L43

Since "full-width" is essentially a width of `100%`, it seems the best/easiest solution would be to set the image to 100% width when the full-width layout is selected. I.e.

```css
.layout-full-width .header-image img {
    width: 100%;
}
```

So before:

<img width="1978" alt="screenshot 2017-06-19 20 01 02" src="https://user-images.githubusercontent.com/52581/27275372-35366cd2-552b-11e7-8213-8aab459e151c.png">

And after:

<img width="2125" alt="screenshot 2017-06-19 20 00 01" src="https://user-images.githubusercontent.com/52581/27275371-35354e88-552b-11e7-814c-664f119308c4.png">

Image quality will obviously take a hit, since the 1480px image will be stretched to whichever width the browser is viewed at. 

I think if the user is going to use the full-width layout, in combination with full-width header images, they'll need to be aware they may want to specify a new header size, finding a sweet spot between file size and image quality. They can do that by using the `themedd_custom_header_args ` filter.

Interested to hear your thoughts!

**Comment by davidsherlock** [comment]
@amdrew, the CSS approach you mention should be fine, but ideally if full-width mode is enabled (via the customizer) conditionally the header image select would disable cropping and allow larger images to be uploaded (for example in the 1920px range). I'm not sure if this is possible or not though?

**Comment by amdrew** [comment]
@davidsherlock I see what you're saying. If full-width mode is enabled, there shouldn't be a restriction on how wide the image is. A user could then upload a 5000px image and then crop it to whichever size they want, say 1920px. I'll have another play.

**Comment by amdrew** [comment]
Ahh ace, totally missed this section where you can specify `flex-width` 👍 
https://codex.wordpress.org/Custom_Headers#Use_flexible_headers

**Comment by amdrew** [comment]
This seems like a better approach. When flex-width headers are specified, the user can skip cropping and upload an image at whichever dimensions they like.

Here's before, note there's only a crop button because Themedd was forcing an image size of 1480x280

<img width="1589" alt="screenshot 2017-06-19 21 11 23" src="https://user-images.githubusercontent.com/52581/27278093-7a90a262-5534-11e7-826e-1cd2b32d31f5.png">

And with flex-width, there's now a "Skip Cropping" button. 1480x280 is only recommended. Best of both words IMO - they can crop an image to 1480 which is recommended by the theme, or if using the full-width layout they can upload any size image they want and skip the cropping.

<img width="1687" alt="screenshot 2017-06-19 21 14 32" src="https://user-images.githubusercontent.com/52581/27278123-92c75e0c-5534-11e7-814e-f479b47e6227.png">

**Comment by davidsherlock** [comment]
Perfect @amdrew! That's way more usable. It's also better because I believe the height won't be forced now as well.

**Comment by amdrew** [comment]
Actually, adding the CSS mentioned before will be a good idea. Although the user can upload any size they now wish, they're going to have to upload something very large to cover the full-width of the browser. Here's a 1920x500px image I uploaded as a test:

<img width="2271" alt="screenshot 2017-06-19 21 24 11" src="https://user-images.githubusercontent.com/52581/27278544-cf090cac-5535-11e7-8adb-ff4b4a7947b2.png">

Since the user choose a "full-width" layout, I think it's safe to say they actually expect this:

<img width="2272" alt="screenshot 2017-06-19 21 26 33" src="https://user-images.githubusercontent.com/52581/27278628-fdcab6e4-5535-11e7-9ecd-764da3aa80aa.png">

There's still nothing wrong with this approach - by letting the user upload any size image, they can make the trade-off between image quality and size.

**Comment by amdrew** [comment]
@davidsherlock Yeah, the unrestricted height is a bonus for sure. This is one of those things that will need to be documented. 

"Why does my image look blurry when the full-width layout is used"
Make sure you have uploaded a large enough image that you're happy with in terms of image quality and file-size.

**Comment by amdrew** [comment]
Previously, the header image was outputted using the `srcset` and `sizes` image attributes. This allows responsive images at different browser widths:
https://github.com/easydigitaldownloads/themedd/blob/master/includes/header.php#L246

With a flex-width header image, the code recommended by WordPress changes as per https://codex.wordpress.org/Custom_Headers#Use_flexible_headers. Basically, the srcset and sizes image attributes are removed.

I think I'll output the image using `srcset` and `sizes` as the default so users can benefit from responsive images and if the full width layout is enabled, I'll use the alternate code recommended.

**Comment by amdrew** [comment]
@davidsherlock Give that latest commit a go.

**Comment by davidsherlock** [comment]
Working well here @amdrew, one suggestion however - when full-width mode is enabled and you crop the image height, the width is still cropped at 1480. Is there anyway to force the width to remain unchanged or a simpler solution would be to utilize the css above in this scenario.

**Comment by davidsherlock** [comment]
wait, it actually does do that, but it dosen't show in real-time in the customizer! ha

**Comment by amdrew** [comment]
Not sure I understand, screenshots?

**Comment by davidsherlock** [comment]
Apologies for the delay in getting back to you on this one @amdrew! I can't actually post screenshots, since the issue only appears on my higher resolution monitor and I'm in a coffee shop on my macbook at the minute working.

To sum up, if you upload an image within the customizer 1480px or larger in width, it dosen't "stretch" in real-time until you save and go back to the site itself. It's more a customizer preview issue to be honest!

**Comment by davidsherlock** [comment]
@amdrew, I think this one can be safely closed now. Full-width mode was added and I haven't seen the above issue re-appear.
